src/java/org/apache/cassandra/index/sai/iterators/KeyRangeConcatIterator.java

```diff
@@ -18,100 +18,105 @@
 package org.apache.cassandra.index.sai.iterators;
 
 import java.util.ArrayList;
-import java.util.Comparator;
 import java.util.List;
-import java.util.PriorityQueue;
 
 import com.google.common.annotations.VisibleForTesting;
 
 import org.apache.cassandra.index.sai.utils.PrimaryKey;
 import org.apache.cassandra.io.util.FileUtils;
 
 /**
- * {@link KeyRangeConcatIterator} takes a list of sorted range iterator and concatenates them, leaving duplicates in
+ * {@link KeyRangeConcatIterator} takes a list of sorted range iterators and concatenates them, leaving duplicates in
  * place, to produce a new stably sorted iterator. Duplicates are eliminated later in
  * {@link org.apache.cassandra.index.sai.plan.StorageAttachedIndexSearcher}
  * as results from multiple SSTable indexes and their respective segments are consumed.
- *
+ * <p>
  * ex. (1, 2, 3) + (3, 3, 4, 5) -> (1, 2, 3, 3, 3, 4, 5)
  * ex. (1, 2, 2, 3) + (3, 4, 4, 6, 6, 7) -> (1, 2, 2, 3, 3, 4, 4, 6, 6, 7)
- *
- * TODO Investigate removing the use of PriorityQueue from this class <a href="https://issues.apache.org/jira/browse/CASSANDRA-18165">CASSANDRA-18165</a>
  */
 public class KeyRangeConcatIterator extends KeyRangeIterator
 {
     public static final String MUST_BE_SORTED_ERROR = "RangeIterator must be sorted, previous max: %s, next min: %s";
-    private final PriorityQueue<KeyRangeIterator> ranges;
-    private final List<KeyRangeIterator> toRelease;
+    private final List<KeyRangeIterator> ranges;
 
-    protected KeyRangeConcatIterator(KeyRangeIterator.Builder.Statistics statistics, PriorityQueue<KeyRangeIterator> ranges)
+    private int current;
+
+    protected KeyRangeConcatIterator(KeyRangeIterator.Builder.Statistics statistics, List<KeyRangeIterator> ranges, Runnable onClose)
     {
-        super(statistics);
+        super(statistics, onClose);
+
+        if (ranges.isEmpty())
+            throw new IllegalArgumentException("Cannot concatenate empty list of ranges");
 
-        this.ranges = ranges;
-        this.toRelease = new ArrayList<>(ranges);
+        this.current = 0;
+        this.ranges = new ArrayList<>(ranges);
```

Comment:
nit: Do we even need the defensive copy? `ranges` is private and never accessible by users of the iterator, and there is already an `ArrayList` created in the builder, which is also private, right?

Reply:
I guess I did not understand correctly this comment:

> If we used my implementation, we could rename toRelease to just ranges I guess?
So you suggest it to be `this.ranges = ranges`? (I just checked that this is what we already do in other iterators too)

Reply:
> So you suggest it to be this.ranges = ranges?

Yeah, basically...the defensive copy just seems paranoid.

Reply:
Done